cmd/nomos/migrate/configmanagement.go

```diff
@@ -0,0 +1,233 @@
+
+	cr := &unstructured.Unstructured{}
+	cr.SetGroupVersionKind(kinds.ClusterRole())
+	cr.SetName(util.ACMOperatorDeployment)
+	objs = append(objs, cr)
+
+	crb := &unstructured.Unstructured{}
+	crb.SetGroupVersionKind(kinds.ClusterRoleBinding())
+	crb.SetName(util.ACMOperatorDeployment)
+	objs = append(objs, crb)
+
+	dep := &unstructured.Unstructured{}
+	dep.SetGroupVersionKind(kinds.Deployment())
+	dep.SetName(util.ACMOperatorDeployment)
+	dep.SetNamespace(configmanagement.ControllerNamespace)
+	objs = append(objs, dep)
+
+	return objs
+}
+
+// executeConfigManagementMigration performs the migration from ConfigManagement
+// installation to a standalone OSS Config Sync installation. It's assumed that
+// the ConfigManagement installation is already running in the multi-repo mode.
+func executeConfigManagementMigration(ctx context.Context, sc *status.ClusterClient) error {
+	printInfo(waitingForRootSyncCRD)
+	if err := waitForMultiRepoCRDsToBeEstablished(ctx, sc.Client); err != nil {
+		return err
+	}
+	printInfo("The Config Sync CRDs have been established")
+
+	printInfo(waitingForReconcilerManager)
+	if err := waitForPodToBeRunning(ctx, sc.K8sClient, configmanagement.ControllerNamespace, "app=reconciler-manager"); err != nil {
+		return err
+	}
+	printInfo("The reconciler-manager Pod is running")
+
+	printInfo(waitingForRGManager)
+	if err := waitForPodToBeRunning(ctx, sc.K8sClient, configmanagement.RGControllerNamespace, "configsync.gke.io/deployment-name=resource-group-controller-manager"); err != nil {
+		return err
+	}
+	printInfo("The resource-group-controller-manager Pod is running")
+
+	// Delete the config-management-operator objects in the reverse order in which
+	// they must be established.
+	printInfo("Deleting the ConfigManagement operator")
+	operatorObjects := getOperatorObjects()
+	for idx := range operatorObjects {
+		obj := operatorObjects[len(operatorObjects)-1-idx]
+		printInfo("deleting object: %s %s/%s",
+			obj.GetObjectKind().GroupVersionKind(), obj.GetNamespace(), obj.GetName())
+		propPolicy := client.PropagationPolicy(metav1.DeletePropagationForeground)
+		if obj.GetObjectKind().GroupVersionKind() == kinds.ConfigManagement() {
+			// The operator adds ownerReferences pointing to the ConfigManagement object,
+			// so "Orphan" must be specified to avoid deleting the managed objects.
+			propPolicy = client.PropagationPolicy(metav1.DeletePropagationOrphan)
+			if err := sc.ConfigManagement.RemoveFinalizers(ctx); err != nil {
+				return err
+			}
+		}
+		if err := sc.Client.Delete(ctx, obj, propPolicy); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return err
+			}
+		}
+		key := client.ObjectKeyFromObject(obj)
+		// wait for object to be not found
+		err := recheck(func() error {
+			err := sc.Client.Get(ctx, key, obj)
+			if apierrors.IsNotFound(err) {
+				return nil
+			}
+			return err
+		})
+		if err != nil {
+			return err
+		}
+		// If this object is the Deployment, wait for all of its Pods to be deleted
```

Comment:
It uses the `foreground` propagation policy, so Pod should have been terminated before the Deployment is gone. Do we need to check again?

Reply:
Surprisingly from my testing it doesn't seem like this waits for the Pods to be terminated. When using `kubectl delete --cascase=foreground` it does seem to wait, I'm not sure why this client doesn't wait.

Reply:
This `Delete` always sets policy to background:
https://github.com/sdowell/kpt-config-sync/blob/814150dd136ed9c92c4fee48cb57e5626ab3c4b7/pkg/syncer/client/client.go#L81-L83
```go
// Delete deletes the given obj from Kubernetes cluster and records prometheus metrics.
// This automatically sets the propagation policy to always be "Background".
func (c *Client) Delete(ctx context.Context, obj client.Object, opts ...client.DeleteOption) status.Error {
```

Reply:
Fixed a bug in the NotFound check above, and removed the specific Pod check